volumina/tiling/cache.py

```diff
@@ -0,0 +1,299 @@
+        if uid not in self._caches:
+            cache = defaultdict(self._default_factory)
+            self._caches[uid] = cache
+        else:
+            raise MultiCacheException("MultiCache.add: uid %s is already in use" % str(uid))
+
+        # remove oldest cache, if necessary
+        self._clean()
+
+    def __contains__(self, key):
+        return key in self._caches
+
+    def __getitem__(self, key):
+        return self._caches[key]
+
+    def __iter__(self):
+        return iter(self._caches)
+
+    def __len__(self):
+        return len(self._caches)
+
+    def touch(self, uid):
+        self._caches.move_to_end(uid)
+
+    @property
+    def maxsize(self):
+        return self._policy.size
+
+    def _evict_one(self):
+        self._caches.popitem(last=False)  # removes item in FIFO order
+
+    def _clean(self):
+        while len(self._caches) > self.maxsize:
+            self._evict_one()
+
+
+class TilesCache:
+    """
+    Contains the following caches, with convenience accessor functions for each.
+
+        layerCache: A cache of 'layers', i.e. for every patch a QImage or QGraphicsItem
+                    for every "image source" in the stack
+
+        tileCache: A cache of 'tiles', i.e. the blended QImage objects
+                   that were created by combining all QImage layers from layerCache for a given patch.
+                   (The QGraphicsItem layers do not contribute to the composite tiles in the tileCache.
+                   They are merely stored.)
+
+        layerCacheDirty: A cache of dirty bits for all layers in layerCache
+        layerCacheTimestamp: A cache of timestamps to track how recently each layer was needed.
+
+        tileCacheDirty: A cache of dirty bits for the composite tiles
+                        (i.e. for a given patch, if a single layer in the patch
+                        is dirty, then the tile for that patch is dirty)
+    """
+
+    def __init__(self, first_stack_id, sims, maxstacks):
+        self._lock = threading.Lock()
+        self._sims = sims
+        self._maxstacks = maxstacks
+        self._policy = CachePolicy(maxstacks)
+
+        kwargs = {"first_uid": first_stack_id, "policy": self._policy}
+
+        # [stack_id][tile_id] -> QImage or QGraphicsItem
+        self._tileCache = MultiCache(default_factory=lambda: (None, 0.0), **kwargs)
+
+        # [stack_id][tile_id] -> bool
+        self._tileCacheDirty = MultiCache(default_factory=lambda: True, **kwargs)
+
+        # [stack_id][(ims, tile_id)] -> QImage or QGraphicsItem
+        self._layerCache = MultiCache(**kwargs)
+
+        # [stack_id][(ims, tile_id)] -> bool
+        self._layerCacheDirty = MultiCache(default_factory=lambda: True, **kwargs)
+
+        # [stack_id][(ims, tile_id)] -> float
+        self._layerCacheTimestamp = MultiCache(default_factory=float, **kwargs)
+
+    @property
+    def maxstacks(self):
+        return self._maxstacks
+
+    def set_maxstacks(self, maxstacks):
```

Comment:
Make this a setter property?

Reply:
I prefer setter method, because Exceptions on setting an attribute usually unexpected behaviour, but in methods it's perfectly normal occurence. I would be conviused seeing
```pyton
try:
   obj.attr = 12
except ValueError:
   pass
```

Reply:
I think it's normal to have an exception when setting an attribute, considering Python's dynamic behaviour. For example, `a = numpy.zeros(4); a.shape = (2, 2)` works, but `a = numpy.zeros(4); a.shape = (2, 999)` raises `ValueError`. You can change an index in pandas dataframe in a similar fashion.

Reply:
I think it's also a matter of style preference. Here setting property also could lead to evictions from the cache. I'll leave it as is.